### DISCUSSION TITLE
[New] Privilégier HTTP/2 à HTTP/1.1

### DISCUSSION THREAD

**Discussion by tbroyer** [cnumr/best-practices#150] [Bonnes pratiques] [answered]
https://groups.google.com/g/ecoconceptionweb/c/bBn3nZ2Ici8/m/gb_KU-TXBgAJ

> “the ingress traffic bandwidth was reduced significantly (close to 50%) after we enabled HTTP/2” — Source: https://blogs.dropbox.com/tech/2016/05/enabling-http2-for-dropbox-web-services-experiences-and-observations/

https://groups.google.com/g/ecoconceptionweb/c/X-ruKr_4fl8/m/kaRQQW7QAgAJ

> HTTP/2 permet de partager une connexion TCP pour plusieurs requêtes (multiplexing, on n'a donc plus besoin que d'une seule connexion par "origine", et au passage un seul handshake HTTPS), et utilise un format binaire plutôt que textuel, avec une table de compression pour les entêtes HTTP (on n'envoie donc plus un entête "Content-Type: text/html", mais un identifiant d'entête qui correspond à "Content-Type" –i.e. un simple octet–, parfois même l'entête avec sa valeur peut être encodé en un seul octet ! et tout le monde partage une table statique prédéfinie des entêtes et valeurs les plus communes, avec une table dynamique liée à la connexion TCP et donc partagée par toutes les requêtes et réponses transmises sur cette connexion). Il y a aussi des fonctionnalités de contrôle de congestion, priorité des flux au sein de la connexion, etc.
Bien entendu, cette table de compression dynamique est maintenue par le client et le serveur, et demande donc un peu plus de ressources (et surtout complexifie beaucoup le code de la stack HTTP/2 par rapport à HTTP/1.x).

> CloudFlare a publié un article […] sur cette compression des entêtes : https://blog.cloudflare.com/hpack-the-silent-killer-feature-of-http-2/
Morceaux choisis :
“When HPACK needs to encode a header in the format name:value, it will first look in the static and dynamic dictionaries. If the full name:value is present, it will simply reference the entry in the dictionary. This will usually take one byte, and in most cases two bytes will suffice! A whole header encoded in a single byte! How crazy is that?»
[…]
“In this case two fields are again present in the static dictionary and five more are repeated and therefore present in the dynamic dictionary, that means they can be encoded in one or two bytes each. One of those is the ~300 byte cookie header, and ~130 byte user-agent. That is 430 bytes encoded into mere 4 bytes, 99% compression!”
[…]
“On average we are seeing a 76% compression for ingress headers. As the headers represent the majority of ingress traffic, it also provides substantial savings in the total ingress traffic: [graphique] We can see that the total ingress traffic is reduced by 53% as the result of HPACK compression!
In fact today, we process about the same number of requests for HTTP/1 and HTTP/2 over HTTPS, yet the ingress traffic for HTTP/2 is only half that of HTTP/1.”
[…]
“The more requests are being processed, the bigger the dynamic table becomes, and more headers can be matched, leading to increased compression ratio.”
[…]
“On average egress headers are compressed by 69%. The savings for the total egress traffic are not that significant however: [graphique] It is difficult to see, but we get 1.4% savings in the total egress HTTP/2 traffic. While it does not look like much, it is still more than increasing the compression level for data would give in many cases. This number is also significantly skewed by websites that serve very large files: we measured savings of well over 15% for some websites.”
[…]
“Conclusion
By implementing HPACK compression for HTTP response headers we've seen a significant drop in egress bandwidth.”
>
> Donc, des gains significatifs en consommation de bande passante, mais en contre-partie une plus grande complexité et une plus grande utilisation mémoire (client et serveur).

Par contre, **ne pas** utiliser les "push promises" de HTTP/2, une fausse bonne idée: https://jakearchibald.com/2017/h2-push-tougher-than-i-thought/ et https://calendar.perfplanet.com/2016/cache-digests-http2-server-push/

En terme de support dans les navigateurs, il est très large depuis plusieurs années: https://caniuse.com/http2
Côté serveurs et CDNs c'est à peu près la même chose: https://istlsfastyet.com/#server-performance

**Comment by DocRoms** [comment]
Ca c'est une très bonne pratique, je pensais qu'on l'avais déjà 😅 .

Par contre, ne vaudrait il pas mieux spécifier `Privligier les protocoles HTTP récents` ? 

`HTTP/3` (HTTP2 QUIC) est déjà bien présent dans l'écosystème (utilisé par FB, ou google). Et améliore encore la transmissions des données. de mémoire, vu que le protocole utilise TLS 1.3, ça permet d'éviter de transmettre des requêtes supplémentaire relatif au chiffrement des données. Néanmoins, vu que c'est de l'UDP, est ce qu'il n'y a pas plus de paquets perdus ? (ensuite, théoriquement, vu que c'est de l'UDP, on a plus l'attente de la confirmation de réception des paquets 🤔 ).

Je ne sais pas si il y a un expert des protocoles HTTP parmi nous, mais je trouve que ça vaut peut-être le coup de se renseigner ? Histoire de rendre cette règle un peu plus générique et que la règle ne deviennent pas obsolète trop rapidement).

Sources: 
 - https://w3techs.com/technologies/details/ce-http3
 - https://gf.dev/http3-test (test de son infra pour HTTP/3)
 - https://www.fasterize.com/fr/blog/protocole-http3/

**Reply by ACTLEM** [reply]
J'ai un petit bémol pour HTTP 3, c'est encore récent et il doit y avoir encore beaucoup de terminaux avec des versions de navigateurs incompatibles (ex: ceux qui utilisent Firefox pour Android surtout depuis la fin du certificat IdentTrust DST Root CA X3 ^^): https://caniuse.com/http3

Sinon, 100% ok pour HTTP2, ça me semble évident aussi.

**Reply by tbroyer** [reply]
Note: c'est un peu pour ça que j'ai précisé de le privilégier "à HTTP/1.1" ; si on met HTTP/3 dans l'équation, il faut commencer par mesurer les gains et éventuels compromis (il y a un an et demi, les performances n'étaient pas meilleures chez Cloudflare, les choses ont probablement bougé depuis, et le test ici était sur la vitesse et pas l'usage de resources: https://blog.cloudflare.com/http-3-vs-http-2/ ) ; au moins formulé en "HTTP/2 vs HTTP/1.1" ça pourra difficilement "mal vieillir".

**Reply by tbroyer** [reply]
À propos d'HTTP/3, un premier retour sur QUIC (un genre de "TLS over UDP" sur lequel est bâti HTTP/3, qui remplace le "TLS over TCP" de HTTP/2) du point de vu CPU: https://www.fastly.com/blog/measuring-quic-vs-tcp-computational-efficiency

**Comment by ACTLEM** [comment, marked as answer]
Note équipe projet: Ok pour la création de la BP

Issue créée: https://github.com/cnumr/best-practices/issues/170
